From c8bc5bbd3ab7ed1850b0e7ef7e7aceacde3fade8 Mon Sep 17 00:00:00 2001
From: Yumi Yukimura <me.cafebabe@gmail.com>
Date: Sat, 6 Jan 2024 00:41:48 +0800
Subject: [PATCH 4/4] build: Add support to attach bootconfig to ramdisk

* Useful for devices which desires to use bootconfig
  (most likely because of running out of cmdline buffer)
  but does not have vendor_boot partition support
* Reference: https://docs.kernel.org/admin-guide/bootconfig.html#attaching-a-boot-config-to-initrd

Change-Id: I0b70d6635dbbff11023eb6c97342c0ea6ad05200
---
 core/Makefile | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/core/Makefile b/core/Makefile
index 47f702a073..eba53c03e9 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -200,6 +200,7 @@ endif
 endif
 
 # This will get referenced by ramdisk(-recovery).img target too
+INTERNAL_VENDOR_BOOTCONFIG_TARGET := $(PRODUCT_OUT)/vendor-bootconfig.img
 INSTALLED_VENDOR_RAMDISK_TARGET := $(PRODUCT_OUT)/vendor_ramdisk.img
 
 include $(BUILD_SYSTEM)/sysprop.mk
@@ -356,13 +357,18 @@ ifeq ($(COMBINING_RAMDISK_IMAGES),true)
   ifeq ($(BUILDING_VENDOR_RAMDISK_IMAGE),true)
     INTERNAL_RAMDISK_IMAGES_COMBINATION_DEPS += $(INSTALLED_VENDOR_RAMDISK_TARGET) $(INTERNAL_VENDOR_RAMDISK_FRAGMENT_TARGETS)
   endif
+  ifdef INTERNAL_BOOTCONFIG
+    INTERNAL_RAMDISK_IMAGES_COMBINATION_DEPS += $(INTERNAL_VENDOR_BOOTCONFIG_TARGET)
+  endif
 endif
 
 # $1: ramdisk image target
 define combine-ramdisk-images
 $(if $(filter true,$(COMBINING_RAMDISK_IMAGES)),\
   $(if $(filter true,$(BUILDING_VENDOR_RAMDISK_IMAGE)),\
-    cat $(INSTALLED_VENDOR_RAMDISK_TARGET) $(INTERNAL_VENDOR_RAMDISK_FRAGMENT_TARGETS) >> $(1)))
+    cat $(INSTALLED_VENDOR_RAMDISK_TARGET) $(INTERNAL_VENDOR_RAMDISK_FRAGMENT_TARGETS) >> $(1))
+  $(if $(INTERNAL_BOOTCONFIG),\
+    prebuilts/tools-lineage/${HOST_OS}-x86/bin/bootconfig -a $(INTERNAL_VENDOR_BOOTCONFIG_TARGET) $(1)))
 endef
 
 # -----------------------------------------------------------------
@@ -1385,6 +1391,15 @@ endif # BOARD_PREBUILT_INIT_BOOT_IMAGE
 
 endif # BUILDING_INIT_BOOT_IMAGE is not true
 
+# -----------------------------------------------------------------
+# vendor bootconfig image
+ifdef INTERNAL_BOOTCONFIG
+$(INTERNAL_VENDOR_BOOTCONFIG_TARGET):
+	rm -f $@
+	$(foreach param,$(INTERNAL_BOOTCONFIG), \
+	 printf "%s\n" $(param) >> $@;)
+endif
+
 # -----------------------------------------------------------------
 # vendor ramdisk image
 INSTALLED_FILES_OUTSIDE_IMAGES := $(filter-out $(TARGET_VENDOR_RAMDISK_OUT)/%, $(INSTALLED_FILES_OUTSIDE_IMAGES))
@@ -1452,11 +1467,6 @@ ifdef INTERNAL_KERNEL_CMDLINE
 endif
 
 ifdef INTERNAL_BOOTCONFIG
-  INTERNAL_VENDOR_BOOTCONFIG_TARGET := $(PRODUCT_OUT)/vendor-bootconfig.img
-  $(INTERNAL_VENDOR_BOOTCONFIG_TARGET):
-	rm -f $@
-	$(foreach param,$(INTERNAL_BOOTCONFIG), \
-	 printf "%s\n" $(param) >> $@;)
   INTERNAL_VENDOR_BOOTIMAGE_ARGS += --vendor_bootconfig $(INTERNAL_VENDOR_BOOTCONFIG_TARGET)
 endif
 
-- 
2.39.2

